[addons][pvr] add independent instance settings way

AlwinEsch opened the pull request:
## Description

The background to this request is to make PVR add-ons usable multiple times.

This is then done using an `instance-settings.xml` file and instance handling adapted in the addon.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

![Bildschirmfoto vom 2021-12-17 04-03-51](https://user-images.githubusercontent.com/6879739/149660394-efef96b4-e449-4a0f-a47f-94dcf496c23e.png)
![Bildschirmfoto vom 2022-01-08 15-31-33](https://user-images.githubusercontent.com/6879739/149660383-2f1cecf7-ec73-4575-bb43-3d14ffa007fa.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Montellese commented on xbmc/settings/lib/SettingSection.h:
```diff
@@ -122,6 +122,7 @@ class CSettingCategory : public ISetting
   bool CanAccess() const;
 
   void AddGroup(const SettingGroupPtr& group);
+  void AddGroupToBegin(const SettingGroupPtr& group);
```
`AddGroupToFront()`

Montellese replied:
Yeah the STL has it all: `begin()` and `end()`, `front()` and `back()` but the most matching name here would be `push_front()` like in `std::list` and other containers. That's why I would to with `AddGroupToFront()`. It also reads more naturally IMO.

Montellese reviewed (dismissed):
I was wondering if it would make more sense to introduce a `CAddonInstance` class which holds the settings of a specific instance to avoid having to add the `instance` parameter to almost all methods in `IAddon` and `CAddon`. Then we would only have to add a `GetAddonInstance(instance)` method. Furthermore `CAddonInstance` could even also implement `IAddon` and could either return itself when `GetAddonInstance(instance)` is called with the proper `instance` or `nullptr` if the `instance` doesn't match.

It might also be a performance improvement in all those places where you have to call multiple methods of `IAddon` and do the instance lookup every time.

AlwinEsch commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -197,9 +197,9 @@ void CPVRClients::UpdateAddons(const std::string& changedAddonId /*= ""*/)
 }
 
-bool CPVRClients::RequestRestart(const std::string& id, bool bDataChanged)
+bool CPVRClients::RequestRestart(const std::string& addonId, uint32_t instance, bool bDataChanged)
```
Yes, was come by change of his parent class, before PVR becomed it, as last commit was the support added to PVR and where it is needed, the `return StopClient(addonId, instance, true);` below..

AlwinEsch commented:
> I was wondering if it would make more sense to introduce a `CAddonInstance` class which holds the settings of a specific instance to avoid having to add the `instance` parameter to almost all methods in `IAddon` and `CAddon`. Then we would only have to add a `GetAddonInstance(instance)` method. Furthermore `CAddonInstance` could even also implement `IAddon` and could either return itself when `GetAddonInstance(instance)` is called with the proper `instance` or `nullptr` if the `instance` doesn't match.
> 
> It might also be a performance improvement in all those places where you have to call multiple methods of `IAddon` and do the instance lookup every time.

Also considered putting it separately and created an `IAddonSettings` class which should then be given by CAddon and IAddonInstanceHandler.

Just thought then to leave it in existing code so related stuff (GUI) has to deal with just one class. In the GUI it is very limited to IAddon in `CGUIDialogAddonSettings::ShowForAddon` and the other functions.
But think about it again, maybe I'll think of something else.

AlwinEsch commented:
Maybe easier possible to manage, if mostly only `CAddonInfo` becomes used and CAddon only on related processes where does his own works.

ksooo commented on xbmc/addons/AddonStatusHandler.cpp:
```diff
@@ -38,8 +38,13 @@ namespace ADDON
+                                         ADDON_STATUS status,
+                                         std::string message,
+                                         bool sameThread)
+  : CThread(("AddonStatus " + std::to_string(instanceID) + "@" + addonID).c_str()),
```
> requested from @ksooo

 instance@addon readsnatural, imo - pretty much like user@host. Initially Alwin came up with addon@instance, which feels like wrong order to me.

ksooo commented on xbmc/addons/Addon.cpp:
```diff
@@ -51,24 +52,30 @@ CAddon::CAddon(const AddonInfoPtr& addonInfo, TYPE addonType)
  */
-bool CAddon::HasSettings()
+
+bool CAddon::CanHaveAddonOrInstanceSettings()
```
The name Alwin originally came up with was not reflecting what the implementation does. But the (quite unusual) name I suggested imo does.

ksooo commented on xbmc/addons/AddonEvents.h:
```diff
@@ -8,14 +8,17 @@
 
 #pragma once
 
-#include <string>
```
Self-contained header files, please. So yes, put it back.

ksooo commented on xbmc/addons/AddonEvents.h:
```diff
@@ -41,6 +44,22 @@ namespace ADDON
     };
 
+    /**
+     * Emitted after a new usable addon instance was added.
```
add-on is correct. In German we use Addon (without the dash).

ksooo commented on xbmc/addons/binary-addons/AddonInstanceHandler.h:
```diff
@@ -24,11 +24,13 @@ class IAddonInstanceHandler
 public:
   IAddonInstanceHandler(ADDON_TYPE type,
                         const AddonInfoPtr& addonInfo,
+                        uint32_t instanceNumber = KODI::ADDONS::ADDON_INSTANCE_ID_UNUSED,
```
> Now we have instance, instanceID and instanceNumber. Please pick one and keep it consistent.

+1 for that

ksooo commented on xbmc/addons/addoninfo/AddonInfo.cpp:
```diff
@@ -144,6 +150,13 @@ TYPE CAddonInfo::TranslateSubContent(const std::string& content)
+{
+  const auto it = std::find_if(types.begin(), types.end(),
+                               [type](const TypeMapping& entry) { return entry.type == type; });
+  return it == types.end() ? AddonInstanceUse::NONE : it->instance_use;
```
Haha, I have no hard feelings about this.

ksooo commented on xbmc/addons/addoninfo/AddonInfo.cpp:
```diff
@@ -243,4 +256,49 @@ const std::string& CAddonInfo::GetTranslatedText(const std::unordered_map<std::s
+  if (!m_supportsInstanceSettings)
+    return singletonInstance;
+
+  const std::string searchPath = StringUtils::Format("special://profile/addon_data/{}/", m_id);
```
> Is this path also used somewhere else? Maybe it can be put into a constant?

@AlwinEsch exactly what I suggested - do you remember?

AlwinEsch replied:
I remember :smile:, thats why I mean to make on separate request. I'm normally not like to have two different parts within one request.

The best is to make this format only one time on one place.
Need to confirm that nothing goes wrong with repo content and not bring in silent problems by change of this.
There are also other parts where maybe can moved to CAddonInfo.

AlwinEsch replied:
In another request becomes changed.

ksooo commented on xbmc/addons/AddonStatusHandler.h:
```diff
@@ -30,7 +30,7 @@ namespace ADDON
                           uint32_t instanceID,
                           ADDON_STATUS status,
-                          std::string message,
+                          const std::string& message,
```
> Was requested by @ksooo

 Really? Ooops.

ksooo replied:
I'm just not a big fan of the "copy and move idiom". This can be discussed back and forth. Do it as you like, but either copy & move or const ref & no move.

AlwinEsch replied:
About the `const std::string& message`, maybe it makes sense to remove complete?
As reason it nowhere becomes set.

AlwinEsch replied:
In another request becomes the unused string removed.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4887,28 +4887,73 @@ msgctxt "#10008"
+#. Header label for add-on configuration removal dialog
+#: xbmc/addons/gui/GUIDialogAddonSettings.cpp
+msgctxt "#10009"
+msgid "Confirm add-on configuration removal"
```
> request of @ksooo

"instance" is far to technical imo. No end user will understand what this means. I consider "configuration" not ideal and I'm totally open for other suggestions.

garbear replied:
The best synonym IMO: "setup". perfectly describes "configuration" and is way simpler. I even created a PR to change 'configuration' to "setup": https://github.com/xbmc/xbmc/pull/11654

Montellese commented on xbmc/addons/binary-addons/AddonInstanceHandler.cpp:
```diff
@@ -22,9 +22,13 @@ CCriticalSection IAddonInstanceHandler::m_cdSec;
 
 IAddonInstanceHandler::IAddonInstanceHandler(ADDON_TYPE type,
                                              const AddonInfoPtr& addonInfo,
+                                             uint32_t instanceNumber /* = ADDON_INSTANCE_ID_U...*/,
```
IMO we shouldn't reduce the length and meaning of variable names and default values just to work around `clang-format`.

AlwinEsch commented:
Is rebased and updated/fixed few parts inside.

AlwinEsch committed:
[addons] increase global API to version 2.0.1 and PVR to 8.1.0

Montellese reviewed (commented):
Unfortunately I don't have the time to look at the changes since my last review.

phunkyfish commented:
Jenkins build this please

ksooo commented:
@AlwinEsch according to the discussion in this PR there are a couple of changes you agreed to make before merging this PR.

AlwinEsch committed:
[addons] Allow CAddonSettings to create with IAddonInstanceHandler

This allows the CAddonSettings class to be created using IAddonInstanceHandler.
This is necessary in order to be able to organize your own settings belonging to the instance.

AlwinEsch committed:
[addons] Add instance change event type

This introduces two new event types into the addon system. These are intended in upcoming commits to inform the respective process system (e.g. upcoming PVR) about them.
The difference to others is that the respective instance id is also transferred.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfo.cpp:
```diff
@@ -144,6 +150,13 @@ TYPE CAddonInfo::TranslateSubContent(const std::string& content)
     return ADDON_UNKNOWN;
 }
 
+AddonInstanceUse CAddonInfo::InstanceUseType(TYPE type)
```
As them as static function not possible, static used to allow also outside.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfo.cpp:
```diff
@@ -12,65 +12,71 @@
 {
-  std::string name;
-  std::string old_name;
+  const char* name;
```
With C++17 it can be becomed used as `std::string_view` and change it to them.

AlwinEsch committed:
[addons] give instance id for settings calls

This is currently limited to the standard settings in settings.xml. As this commit in order to keep it clearer in the following, where its processing is introduced.

AlwinEsch commented:
Is updated and should now includes most changes about the notes, thanks a lot about the improvement tips!

Has done now fast on rebase and still need to check the add-ons about.

AlwinEsch commented:
The first commits relates to https://github.com/xbmc/xbmc/pull/21739 and will go away from here.

AlwinEsch committed:
[addons] give instance ID to CAddonStatusHandler

With this commit, the instance identification is transferred to the call in order to retrieve the correct settings and add-on information.

AlwinEsch committed:
[addons][pvr] Implement multiple instance usage to PVR

With all previous changes it is now possible to support multiple instances in PVR where each can have its own settings.

AlwinEsch committed:
[addons] Allow use of instance settings

This allows the add-on interface to read instance-related settings (if supported by the addon).

In the GUI it cannot be edited yet and comes in the following commits.

AlwinEsch committed:
[addons] Add selection of instance-settings.xml in CAddonInfo

With this, when the CAddonInfo class is created, the availability of the instance settings is checked and made available to the CAddon.

AlwinEsch commented:
@ksooo @Montellese as the other was become in, is them here cleaned up and ready for review and tests.

@ksooo Maybe you can test with your pvr backends, have for now only tested with https://github.com/kodi-pvr/pvr.demo/pull/123

AlwinEsch commented on xbmc/addons/Addon.h:
```diff
@@ -258,54 +398,102 @@ class CAddon : public IAddon
+   * \sa LoadUserSettings, SaveSettings, HasSettings, HasUserSettings, GetSetting, UpdateSetting
    */
-  bool LoadSettings(bool bForce, bool loadUserSettings = true);
+  bool LoadSettings(bool bForce, bool loadUserSettings, AddonInstanceId id = ADDON_SETTINGS_ID);
```
Think to add about here on another following request an change of bool to enum, like now a bit confusing and can be maybe by mistake become id instead of bool.

E.g.:
```cpp
enum class SettingsForceLoad
{
  YES,
  NO
};

enum class SettingsLoadUserPart
{
  YES,
  NO
};

bool LoadSettings(SettingsForceLoad bForce, SettingsLoadUserPart loadUserSettings, AddonInstanceId id = ADDON_SETTINGS_ID);
```

@ksooo What you think about?

enen92 replied:
Better to avoid YES and NO as enum values as they seem to already exist in Darwin and have been a cause for past issues. @howie-f should know better (I think CHOICE_YES and CHOICE_NO has been used instead)

AlwinEsch replied:
Also relates to an "enum class"?

enen92 replied:
Yes, see https://github.com/xbmc/xbmc/pull/20925

ksooo replied:
@AlwinEsch generally I like the idea, but may I ask you not to add more stuff to this PR, please? It is already large and difficult to review. So, I would appreciate if you put this kind of cosmetic enhancements not needed for the addon-API changes in a follow-up PR. For now, let's focus on getting this PR in shape, without adding more to it.

AlwinEsch replied:
Yes, not thought about for here (see the "here on another following request" above), only as the place visible here I thought to ask about 😀.

ksooo replied:
There is much more room for improving several function signatures, like getting rid of bool parameters, replacing them with speaking enum values etc. But all this is really for another PR.

ksooo commented:
> @ksooo Maybe you can test with your pvr backends, have for now only tested with https://github.com/kodi-pvr/pvr.demo/pull/123

Regression tests ensuring ols functionality is not broken can be done by anybody. To test the new functionality, I would need to do larger changes to at least one pvr add-on and I'm afraid I will not find the time for that anytime soon.

ksooo commented on xbmc/pvr/addons/PVRClients.cpp:
```diff
@@ -33,15 +33,15 @@ using namespace PVR;
+{
+  std::hash<std::string> hasher;
+  int iClientId =
+      static_cast<int>(hasher((instance > 0 ? "@" + std::to_string(instance) : "") + addonID));
```
Shouldn't the string to hash read "instance@addonID"? Current implementation produces for non-empty instance "@instanceaddonID"...

ksooo commented on xbmc/pvr/channels/PVRChannelsPath.cpp:
```diff
@@ -81,7 +92,8 @@ CPVRChannelsPath::CPVRChannelsPath(const std::string& strPath)
         {
-          m_kind = Kind::CHANNEL; // pvr://channels/(tv|radio)/<groupname>/<addonid>_<channeluid>.pvr
+          m_kind = Kind::
+              CHANNEL; // pvr://channels/(tv|radio)/<groupname>/[<instanceid>@]<addonid>_<channeluid>.pvr
```
You implemented the new path so that instanceid is always present. So, the comment should read: "// pvr://channels/(tv|radio)/<groupname>/<instanceid>@<addonid>_<channeluid>.pvr"

ksooo reviewed (commented):
minors only

AlwinEsch commented:
Can try to change pvr.zattoo about, as this already allow different web hosts in his settings.

ksooo commented:
I adapted pvr.hts (PR to follow). After ironing out some glitches in this PR (now in), the new feature works like a charm. 

I will merge the PR once Jenkins is green.

BTW: This is all compatible Add- on API changes, so now new add-on release needed.

ksooo reviewed (approved):
A very cool feature, well implemented. :-)

phunkyfish commented:
Excellent, great to see this going in. Now just to find the time to do the same for pvr.vuplus and pvr.iptvsimple! Thanks @AlwinEsch !!!

gade01 commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4901,28 +4903,73 @@ msgctxt "#10008"
+#. Header label for add-on configuration removal selection dialog
+#: xbmc/addons/gui/GUIDialogAddonSettings.cpp
+msgctxt "#10010"
+msgid "Select add-on configuration to remove"
```
@AlwinEsch I k ow this PR already went in, but there's a missing `mscstr ""` at line 4916.

ksooo replied:
I will open a PR with a fix.
